File: crates/db/src/models/gallery.rs
use chrono::NaiveDateTime;
use diesel::pg::Pg;

use crate::{gallery_images, Package};

/// A gallery image.
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Serialize,
    Deserialize,
    Identifiable,
    Queryable,
    Selectable,
    Associations,
    ToSchema,
    ToResponse,
)]
#[diesel(table_name = gallery_images)]
#[diesel(belongs_to(Package, foreign_key = package))]
#[diesel(check_for_backend(Pg))]
pub struct GalleryImage {
    /// The gallery image ID.
    pub id: i32,

    /// The package ID.
    pub package: i32,

    /// The display name of the version.
    pub name: String,

    /// This image's ID in S3.
    pub s3_id: String,

    /// An optional markdown-formatted description.
    pub description: Option<String>,

    /// The order of this image.
    pub ordering: i32,

    /// The date this version was created.
    pub created_at: NaiveDateTime,

    /// The date this version was last updated.
    pub updated_at: NaiveDateTime,
}

/// A gallery image for insertion.
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Serialize,
    Deserialize,
    Queryable,
    Selectable,
    Associations,
    Insertable,
    ToSchema,
    ToResponse,
)]
#[diesel(table_name = gallery_images)]
#[diesel(belongs_to(Package, foreign_key = package))]
#[diesel(check_for_backend(Pg))]
pub struct NewGalleryImage {
    /// The package ID.
    pub package: i32,

    /// The display name of the version.
    pub name: String,

    /// This image's ID in S3.
    pub s3_id: String,

    /// An optional markdown-formatted description.
    pub description: Option<String>,

    /// The order of this image.
    pub ordering: i32,
}

File: ui/src/lib/types/gallery.ts
export interface PublicGalleryImage {
    /**
     * The gallery image ID.
     */
    id: number;

    /**
     * The package ID.
     */
    package: number;

    /**
     * The display name of the version.
     */
    name: string;

    /**
     * A URL to access this image with.
     */
    url: string;

    /**
     * An optional markdown-formatted description.
     */
    description?: string;

    /**
     * The order of this image.
     */
    ordering: number;

    /**
     * A stringified {@link Date} for when this image was created.
     */
    created_at: string;

    /**
     * A stringified {@link Date} for when this image was last updated.
     */
    updated_at: string;
}

export interface GalleryImageInit {
    /**
     * The display name of the version.
     */
    name: string;

    /**
     * An optional markdown-formatted description.
     */
    description?: string;

    /**
     * The order of this image. Defaults to -1.
     */
    ordering?: number;
}

export type GalleryImageUpdate = Partial<GalleryImageInit>;
